### PR TITLE
feat: add content-length headers to generated responses, via new `text(...)` helper

### LINKED ISSUE #5749: Way to access the response body size in the hook
### Describe the problem

I want to set up logging in my sveltekit hook. One of the things I want to log is the number of bytes in the response body. It would be nice if the `Response` exposed the body length so that we can access it. Something like this...

```js
export async function handle({ event, resolve }) {
    response = await resolve(event)
    // these don't actually work, they are undefined. But something LIKE this would be nice
    console.log(response.body.length)
    console.log(response.headers.get("content-length"))
    return response
}
```

I notice that the length is somewhere on the object if I just `console.log(response)` when returning JSON...

```
Response {
  [Symbol(realm)]: { settingsObject: {} },
  [Symbol(state)]: {
    aborted: false,
    rangeRequested: false,
    timingAllowPassed: false,
    requestIncludesCredentials: false,
    type: 'default',
    status: 200,
    timingInfo: null,
    cacheState: '',
    statusText: '',
    headersList: HeadersList {
      [Symbol(headers map)]: [Map],
      [Symbol(headers map sorted)]: null
    },
    urlList: [],
    body: {
      stream: undefined,
      source: '{"result":{"rows":[["to become more autonomous in data analysis"],["to better structure your thoughts"],["to develop a useful professional skill"]],"columns":[{"name":"reason","type":"VARCHAR"}],"moreRows":false}}',
      length: 213
    }
  },
  [Symbol(headers)]: HeadersList {
    [Symbol(headers map)]: Map(2) {
      'content-type' => 'application/json; charset=utf-8',
      'etag' => '"1ba7x9x"'
    },
    [Symbol(headers map sorted)]: null
  }
}
```

...but that piece of data is private (attached to a symbol) and I can't access it.

In my opinion, being able to observe information like this in my hook is important to building debuggable, observable web services. Response payload size is an important metric to monitor. Putting that logging logic in the hook seems to be the only way to make sure it is reliably logged.

### Describe the proposed solution

The workaround I have right now is `await response.clone().text()`, but that is not ideal because I have to consume the stream. This is wasteful especially if I only want the length.

### Alternatives considered

I suppose I could do something with `Object.getOwnPropertySymbols()`, but I think that's frowned upon?

### Importance

would make my life easier

### Additional Information

As I understand it (may be wrong), the content length isn't necessarily always known without reading the stream, but there are many cases where the length is known at the time of the response instantiation, like when JSON or HTML are being returned from a string in memory. Sveltekit should be able to set the length in cases where length is known.

Here is a simple case where content length is known but not exposed: https://stackblitz.com/edit/sveltejs-kit-template-default-9ydsbu?file=src%2Fhooks.js&terminal=dev

### PR DESCRIPTION
**Pull request by Rich-Harris**
closes #5749.

This is just an idea; I don't know if it's the right one. In order to make it easier to determine response content length inside `handle`, for logging purposes, this adds a new `text(body)` helper similar to the `json(data)` helper. It returns a `Response` object with a `content-length` header, and uses it in favour of `new Response(...)` when generating pages and other responses.

One observation: it'd be nice if we could unit-test `Server` rather than having Playwright tests for the bulk of the tests in `server.test.js`. Might look into that separately.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

File: packages/kit/src/exports/index.js [hunk 1]
```diff
@@ -29,12 +29,23 @@ export function redirect(status, location) {
...
+	return response;
+}
+
+/** @type {import('@sveltejs/kit').text} */
+export function text(body, init) {
 	const headers = new Headers(init?.headers);
-	if (!headers.has('content-type')) {
-		headers.set('content-type', 'application/json');
+	if (!headers.has('content-length')) {
+		headers.set('content-length', String(body.length));
```
Comment (bjon) [inline packages/kit/src/exports/index.js:45]: This doesn't work right? Probably only works with ascii.

I had to do `Buffer.byteLength(str).toString()`

Comment (Rich-Harris) [reply packages/kit/src/exports/index.js]: good catch, thank you. the non-Node version of this is `new TextEncoder().encode(body).byteLength.toString()`

File: packages/kit/types/index.d.ts [hunk 2]
```diff
@@ -1122,10 +1122,17 @@ export interface Redirect {
...
  * Create a JSON `Response` object from the supplied data.
  * @param data The value that will be serialized as JSON.
- * @param init Options such as `status` and `headers` that will be added to the response. A `Content-Type: application/json` header will be added automatically.
+ * @param init Options such as `status` and `headers` that will be added to the response. `Content-Type: application/json` and `Content-Length` headers will be added automatically.
  */
 export function json(data: any, init?: ResponseInit): Response;
 
+/**
+ * Create a `Response` object from the supplied body.
+ * @param body The value that will be serialized as JSON.
```
Comment (dummdidumm) [inline packages/kit/types/index.d.ts:1131]:
```suggestion
 * @param body The value that will be used as-is.
```

**Review by dummdidumm — APPROVED** [review]
This approach makes sense to me

**Review by dummdidumm — CHANGES_REQUESTED** [review]
Ok maybe it doesn't make sense..? [This stackoverflow answer](https://stackoverflow.com/questions/50825528/content-length-header-not-the-same-as-when-manually-calculating-it) makes me kinda nervous - it seems that compression can alter the content length, and I'm not sure what bugs we might introduce if we set the uncompressed size

**Comment by Rich-Harris** [comment]
Ah, that's a good point. I would have assumed that if something in between SvelteKit and the user were to compress the body it would generate a new response with new headers, but I'll freely confess my ignorance here. Safest thing would be to close #5749 as wontfix but that seems unsatisfying.

Am inclined to just leave this here until someone with more authority weighs in

**Comment by bjon** [comment]
@dummdidumm Google's dynamic compression (CDN) requires the response to have `content-length`. Without it, no compression is done. https://cloud.google.com/cdn/docs/dynamic-compression#response-not-compressed

**Comment by Conduitry** [comment]
If there's some library or something that compresses responses on the fly but doesn't correctly update the `content-length` header, that's certainly a bug in that library. I think we should only worry about that if there were some extremely popular library that did that that seemed extremely unlikely to want to fix that bug.

**Comment by Rich-Harris** [comment]
I don't think there's much value in having an `addContentLength` option — it doesn't cost anything extra for us, and I can't imagine any cases where you _wouldn't_ want the header. Honestly I'm a little surprised it doesn't get added automatically (if you do `new Response('hello')` it'll add a `content-type: text/plain;charset=UTF-8` header, but not `content-length` — what the hell?)

So it sounds like the consensus is in favour of merging this?

**Review by dummdidumm — APPROVED** [review]
I agree. Rereading the posted Stackoverflow answer, I noticed that the concern there is the other way around - someone's reading the response text and comparing that with the length. So this is the other end, consumer (user in the browser for our case), not producer (our server).

That said - any way we can merge another maybe bit nicer feature before releasing this? 😄 Having 1.1 just for this feels wasted. Not sure how much we care about minor version marketing at this point.

File: .changeset/nine-walls-listen.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@sveltejs/kit': patch
+---
+
+Add `Content-Length` header to SvelteKit-generated responses
```
Comment (benmccann) [inline .changeset/nine-walls-listen.md:5]:
```suggestion
'@sveltejs/kit': minor
---

feat: add `Content-Length` header to SvelteKit-generated responses
```

File: .changeset/old-weeks-reflect.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@sveltejs/kit': minor
+---
+
+Add `text(...)` helper for generating text responses
```
Comment (benmccann) [inline .changeset/old-weeks-reflect.md:5]:
```suggestion
'@sveltejs/kit': minor
---

feat: add `text(...)` helper for generating text responses
```

File: .changeset/nine-walls-listen.md (hunk 3 above)
Comment (benmccann) [inline .changeset/nine-walls-listen.md:5]:
```suggestion
'@sveltejs/kit': patch
---

fix: add `Content-Length` header to SvelteKit-generated responses
```

**Pull request merged**